runestone/common/runestonedirective.py

```diff
@@ -241,7 +248,7 @@ def getNumber(self):
 
         env.assesscounter += 1
 
-        res = "Q-%d"
+        res = "Q\ :runestone_secnum:`%d`"
```

Comment:
I think this will just end up with the question in the rendered html looking like:

Q :runestone_secnum:3

Unless that :runestone_secnum: gets inserted *before* we recursively process the text of the question. But I think it might be too late as we are just dumping this into options["qnumber"] which gets handled when we make the html.  Or not at all for many directives. 

Also if we go this way then there are lots of changes that will need to be made in dbchapterinfo.py where we populate the chapters and sub_chapters tables.

Reply:
This works fine for mchoice and fitb questions, since both of those run the reST parser on the text inside the question. I don't know about the other question types; I look at Parson's problems, and that dumps the question text directly into HTML (is this ever checked for XSS?)

I agree that the qnumber won't be correct, but it would be possible to fix that up during the doctree-resolve phase.

Mainly, I'm unhappy with the idea of yet another tie to the chapter/subchapter structure, since this diverges from what Sphinx does in terns of numbering and is something we know we'll have to fix at some point. This PR was a stab at getting us one step closer to aligning with Sphinx.

Reply:
Why create a new directive (:runestone_secnum:)? Is this ever something that an author would write?  It seems like we are just adding in this new directive to be processed later, when we could just add the number directly to the question.

I think we are setting ourselves up for more work by conflating the following two goals:

1. Have the labels on questions match the section numbers generated by sphinx in the table of contents, and make those display correctly in the assignment building interface.
2. Process the question description for each question recursively with sphinx.

To be clear I am in favor of both goals.

We (all of us at the hack week at Luther) made the decision to follow this chapter/subchapter structure years ago and it works just fine for those that color inside those lines.  I realize that is not how you want to structure your book, but it doesn't help to complain that something is broken when you are using it in a way that it wasn't intended to be used.

It seems to me that the only way out of the sort of technical debt that we have incurred with our chapter/subchapter decision is to come up with a long term solution and then work incrementally toward that. Making things incrementally better and making compromises for compatibility as we go.  The large user base that I have to support on runestone.academy makes it hard to make wholesale breaking changes that will mess up people in the midst of teaching a course.

Reply:
Thanks for the feedback! I always appreciate other perspectives to help refine and improve my ideas.

My goal aligns with your: look for a long-term solution to paying of the chapter/subchapter technical debt. My strategy for accomplishing it one small step at a time is to gradually transition from chapter/subchapter numbering to Sphinx section numbering; if books (unlike mine) follow the Runestone rules, these are identical. Since we need to add code to number questions, this seems like a good point to use Sphinx section numbers as a part of this overall strategy. However, this does means conflating items 1 and 2, which may be too much to bite off at one time. What are your thoughts on the best way to accomplish this long-term solution?

My implementation was based on this strategy: use Sphinx section numbers as the prefix for question numbers. Since Sphinx section numbers aren't available during parsing, I created a role (which would only be used internally) which can be resolved when the section numbers are available. Again, I'm open to other approaches and can abandon this if you see this as moving us away from this long-term goal. I'm also happy to chat a bit more if you'd like to discuss the approach.

In terms of DAL, it seems to me that having developers know just DAL, instead of both SQLAlchemy and DAL, to develop would simplify things -- the components and the server could share code for some of the tables, queries would be similar, etc. For the same reason, it makes sense to me to do a server rewrite using DAL, not SQLAlchemy. However, this is more a long-term suggestion as well -- ways to make the code easier to maintain in the future.

Reply:
My initial thought at where to go with this was to use the sphinx section numbers to populate the chapter and subchapter tables with numbers, ensuring a match.  Then to use the qnumber field of the question table to store the label that we want to display.  That may be flawed for sub-subchapters, more investigation is needed.  That label could then be plugged into the question text when we are walking the nodes and know the numbering for everything.

It is important to have the question label information in the database so that we can use it in the assignment builder and the assignment builder matches the book pages and the table of contents.

We also need to think about backward compatibility for some books that right now do make use of the qnum directive to set a prefix/postfix and plug in a formatted label for the question today.  That required a lot of discipline on the part of the author, nevertheless some authors have used that fairly carefully.  So my thinking was that the qnum directive would override the default of using the sphinx section numbers + question number for the default.

We also need to hash out where these numbers should appear.  Should they be at the beginning of the question text?  Personally I like them a lot better where they appear today, which is below the component.  But I'd like @barbarer and @bhoffman0 and @pearcej to weigh in if they have strong opinions about this.

Reply:
On thing I was thinking about with respect to sub-sub sections. was whether or not we out to think about the book being composed of chapters and pages instead of chapters and subchapters.

I'm not sure how you have structured your book.  Does a sububchapter always live in its own file?  Or can you have subsubchapters on the same page?  If they are on the same page then we need a way to figure out how to number the questions that respects those subsub boundaries.  We would not want to have

Question 2.3.11.1
Question 2.3.11.2
then
Question 2.3.12.3
- what happened to 2.3.12.1 and 2?

Should we just use the chapter and subchapter as part of the question number ignoring sub subchapters In question numbering?

Reply:
With respect to parsing the question text.   I wonder if it would be possible to have every question use the convention in activecode?  Question text followed by a line with `~~~~`. This would make it easier than handling some questions that use `:question:` for the question.

This would let us code this once rather than handling it in each directive.

Reply:
I'm certainly in favor of using Sphinx section numbering wherever it's easy to gather for Runestone. That's certainly a step in the right direction.

What is a question label? Is it a question number, like question 2.3.11.1, or something else? As far as I know, question numbers currently appear at the beginning of a question, while the div_id (label?) appears at the bottom of the question.

My sub-subchapters are always separate files, due to the requirements of the literate programming question type I use.

I think using a consistent method through all the Runestone components, such as dividing question text from other parts with a `~~~~`, is a great idea! However, doesn't it also mean we'll need to update everything question in our books to follow this new format? That seems painful.

Reply:
Some authors use the `..qnum::` directive on each page.

```
.. qnum:: 
   :prefix: func-
   :start: 1
```

For this page every question is labelled/numbered func-X

The feedback from the majority of teachers was that they found that confusing.  So I turned it off completely for fopp.

They seem to like the `Activity X:  Multiple Choice (div_id_here)` that appears below the question.  -- Those are numbered using a simple CSS counter on each page.  My assumption is that they would equally like
`Activity 2.3.4: Multiple Choice (div_id_here)`. Especially if 2.3.4 also appeared in the assignment builder.

I agree that any modification to the question format would be painful, unless it was easy to automate updating all of our book source.  which in this case I doubt.   It turns out that for purposes of the description it is easy enough to just grab the content up to the first blank line for each question, or to override that just for the activecode and clickablearea questions.

Reply:
That sounds good to me. I don't use `qnum` and like the idea of auto-numbered questions as you propose.

Do you want me to do anything on this, or would you like to code it up?